lib/Sema/ConstraintSystem/LocalCSWalker.cpp

```diff
@@ -228,9 +229,77 @@ class LocalCSWalker : public glu::ast::ASTWalker<LocalCSWalker, void> {
         _cs.addConstraint(constraint);
     }
 
-    void _visitForStmt([[maybe_unused]] glu::ast::ForStmt *node)
+    void _visitForStmt(glu::ast::ForStmt *node)
     {
-        // TODO: #439 Implement for-loops
+        auto *binding = node->getBinding();
+        auto *range = node->getRange();
+
+        visit(range);
+
+        auto &typesArena = _astContext->getTypesMemoryArena();
+
+        auto *bindingType = binding ? binding->getType() : nullptr;
+        if (!bindingType) {
+            bindingType = typesArena.create<glu::types::TypeVariableTy>();
+            binding->setType(bindingType);
+        }
+        if (auto *typeVar
+            = llvm::dyn_cast<glu::types::TypeVariableTy>(bindingType)) {
+            _cs.addTypeVariable(typeVar);
+        }
+
+        auto *iteratorType = typesArena.create<glu::types::TypeVariableTy>();
+        _cs.addTypeVariable(iteratorType);
+
+        auto *rangeType = range->getType();
+
+        auto createHelperRef
+            = [&](llvm::StringRef name) -> glu::ast::RefExpr * {
+            auto *ref
+                = _astContext->getASTMemoryArena().create<glu::ast::RefExpr>(
+                    node->getLocation(),
+                    glu::ast::NamespaceIdentifier {
+                        llvm::ArrayRef<llvm::StringRef>(), name },
+                    nullptr
+                );
+            preVisitExprBase(ref);
+            visit(ref);
+            return ref;
+        };
+
+        auto addHelperConstraint =
+            [&](glu::ast::RefExpr *ref,
+                llvm::ArrayRef<glu::types::TypeBase *> params,
+                glu::types::TypeBase *result) {
+                auto *fnTy
+                    = typesArena.create<glu::types::FunctionTy>(params, result);
+                _cs.addConstraint(
+                    Constraint::createConversion(_cs.getAllocator(), ref, fnTy)
+                );
+            };
+
+        auto *beginRef = createHelperRef("begin");
+        node->setBeginFunc(beginRef);
```

Comment:
you need to set it before visiting it, parent should never be nullptr within the CS